File: Linear/Dynamic/Linked List/Singly Linked List/03_Deleting/End/Deleting.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Function to remove the last node
def deleteLastNode(head):
    """Check if the list is empty or If the
    list has only one node, delete it"""
    if head is None or head.next is None:
        return None
    
    # Find the second last node
    secondLast = head
    while secondLast.next.next is not None:
        secondLast = secondLast.next
    
    # Delete the last node
    secondLast.next = None

    return head

# Function to print the linked list
def printList(head):
    while head is not None:
        print(head.data, end="")
        if head.next is not None:
            print(" -> ", end="")
        head = head.next
    print()

# Create the linked list: 1 -> 2 -> 3 -> 4 -> 5
head = Node(1)
head.next = Node(2)
head.next.next = Node(3)
head.next.next.next = Node(4)
head.next.next.next.next = Node(5)

print("Linked List before deletion\n")
printList(head)

head = deleteLastNode(head)

print("Linked List after deletion\n")
printList(head)

File: Linear/Dynamic/Linked List/Singly Linked List/03_Deleting/End/Deleting.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class DeletingEnd {
    // Function to remove the last node
    public static Node deleteLastNode(Node head) {
        /*Check if the list is empty or If the
        list has only one node, delete it*/
        if (head == null || head.next == null)
        return null;

        // Find the second last node
        Node secondLast = head;
        while (secondLast.next.next != null) {
            secondLast = secondLast.next;
        }

        // Delete the last node
        secondLast.next = null;
        
        return head;
    }

    // Function to print the linked list
    public static void printList(Node head) {
        while (head != null) {
            System.out.print(head.data);
            if (head.next != null)
            System.out.print(" -> ");
            head = head.next;
        }

        System.out.println();
    }

    public static void main(String[] args) {
        // Create the linked list: 1 -> 2 -> 3 -> 4 -> 5
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);
        head.next.next.next.next = new Node(5);
        
        System.out.println("Linked List before deletion");
        printList(head);

        head = deleteLastNode(head);
        
        System.out.println("Linked List after deletion");
        printList(head);
    }
}